lib/Dialect/Arc/Transforms/FindInitialVectors.cpp

```diff
@@ -0,0 +1,254 @@
+#include "mlir/IR/Types.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/raw_ostream.h"
+
+#define DEBUG_TYPE "find-initial-vectors"
+
+namespace circt {
+namespace arc {
+#define GEN_PASS_DEF_FINDINITIALVECTORS
+#include "circt/Dialect/Arc/ArcPasses.h.inc"
+} // namespace arc
+} // namespace circt
+
+using namespace circt;
+using namespace arc;
+using llvm::SmallMapVector;
+using llvm::SmallVector;
+using mlir::Operation;
+
+namespace {
+struct TopologicalOrder {
+  /// An integer rank assigned to each operation.
+  SmallMapVector<Operation *, unsigned, 32> opRanks;
+  LogicalResult compute(Block *block);
+  unsigned get(Operation *op) const {
+    const auto *it = opRanks.find(op);
+    assert(it != opRanks.end() && "op has no rank");
+    return it->second;
+  }
+};
+} // namespace
+
+/// Assign each operation in the given block a topological rank. Stateful
+/// elements are assigned rank 0. All other operations receive the maximum rank
+/// of their users, plus one.
+LogicalResult TopologicalOrder::compute(Block *block) {
+  LLVM_DEBUG(llvm::dbgs() << "- Computing topological order in block " << block
+                          << "\n");
+  struct WorklistItem {
+    WorklistItem(Operation *op) : userIt(op->user_begin()) {}
+    Operation::user_iterator userIt;
+    unsigned rank = 0;
+  };
+  SmallMapVector<Operation *, WorklistItem, 16> worklist;
+  for (auto &op : *block) {
+    if (opRanks.contains(&op))
+      continue;
+    worklist.insert({&op, WorklistItem(&op)});
+    while (!worklist.empty()) {
+      auto &[op, item] = worklist.back();
+      if (auto stateOp = dyn_cast<StateOp>(op)) {
+        if (stateOp.getLatency() > 0)
+          item.userIt = op->user_end();
+      } else if (auto writeOp = dyn_cast<MemoryWritePortOp>(op)) {
+        item.userIt = op->user_end();
+      }
+      if (item.userIt == op->user_end()) {
+        opRanks.insert({op, item.rank});
+        worklist.pop_back();
+        continue;
+      }
+      if (auto *rankIt = opRanks.find(*item.userIt); rankIt != opRanks.end()) {
+        item.rank = std::max(item.rank, rankIt->second + 1);
+        ++item.userIt;
+        continue;
+      }
+      if (!worklist.insert({*item.userIt, WorklistItem(*item.userIt)}).second)
+        return op->emitError("dependency cycle");
+    }
+  }
+  return success();
+}
+
+namespace {
+using Key = std::tuple<unsigned, StringRef, SmallVector<Type>,
+                       SmallVector<Type>, ArrayRef<NamedAttribute>>;
+
+Key computeKey(Operation *op, unsigned rank) {
+  // The key = concat(op_rank, op_name, op_operands_types, op_result_types,
+  //                  op_attrs)
+  return std::make_tuple(
+      rank, op->getName().getStringRef(),
+      SmallVector<Type>(op->operand_type_begin(), op->operand_type_end()),
+      SmallVector<Type>(op->result_type_begin(), op->result_type_end()),
+      op->getAttrs());
+}
```

Comment:
I think you can directly use the attribute dictionary instead of an array of named attributes. That makes the key cheaper to compare: the dictionary is just a pointer (interned in the MLIRContext), so it's just a pointer comparison:
```suggestion
using Key = std::tuple<unsigned, StringRef, SmallVector<Type>,
                       SmallVector<Type>, DictionaryAttr>;

Key computeKey(Operation *op, unsigned rank) {
  // The key = concat(op_rank, op_name, op_operands_types, op_result_types,
  //                  op_attrs)
  return std::make_tuple(
      rank, op->getName().getStringRef(),
      SmallVector<Type>(op->operand_type_begin(), op->operand_type_end()),
      SmallVector<Type>(op->result_type_begin(), op->result_type_end()),
      op->getAttrDictionary());
}
```